modules/ting_field_search/ting_field_search.module

```diff
@@ -432,39 +433,202 @@ function ting_field_search_ting_search_settings_submit($form, &$form_state) {
+      if ($visibility) {
+        $pages .= "\r\nsearch-blank";
+      }
+
+      // Compare the lowercase internal and lowercase path alias (if any).
+      $page_match = drupal_match_path($path, $pages);
+      if ($path != $_GET['q']) {
+        $page_match = $page_match || drupal_match_path($_GET['q'], $pages);
+      }
+
+      // When $visibility has a value of 0, the profile is displayed on
+      // all pages except those listed in $pages. When set to 1, it
+      // is displayed only on those pages listed in $pages.
+      if (!($visibility xor $page_match)) {
+        $default_profile = $profile->name;
+        break;
+      }
+    }
   }
 
+  return $default_profile;
+}
+
+/**
+ * Filter the profiles displayed based on configuration.
+ */
+function ting_field_search_filter_profiles() {
   $exposed_profiles = array();
+
+  $active_profile = ting_field_search_get_active_profile();
+
   $profiles = ting_field_search_profiles_load_all();
   foreach (ting_field_search_profiles_sort($profiles) as $name => $profile) {
     // Only exposed and activated profiles should be shown to the user.
     if ($profile->config['user_interaction']['exposed'] && empty($profile->disabled)) {
-      $exposed_profiles[$name] = t('Search in @profile', array(
-       '@profile'  => check_plain($profile->title),
-      ));
+      // Filter based on grouping (always show profiles from same group).
+      if ($active_profile && $active_profile->config['user_interaction']['grouping'] == $profile->config['user_interaction']['grouping']) {
+        $exposed_profiles[$name] = t('Search in @profile', array(
+          '@profile'  => check_plain($profile->title),
+        ));
+        continue;
+      }
+
+      // If the visibility is "expect" and we have a active profile that's
+      // different that the current one (checked above). It should not be shown.
+      if ($profile->config['user_interaction']['visibility'] == 'expect' && $active_profile) {
+        continue;
+      }
+
+      // Filter base on path.
+      if ($profile->config['user_interaction']['visibility'] !== 'all') {
+        $visibility = $profile->config['user_interaction']['visibility'] == 'expect' ? 0 : 1;
+        $pages = $profile->config['user_interaction']['pages'];
+        $path = drupal_strtolower(drupal_get_path_alias($_GET['q']));
+
+        // We have an optimization hack to speed up search (by-pass panels 2x
+        // times loading). So all searches are submitted to "/search-blank",
+        // which need to be added to pages to ensure no form validation errors.
+        // If visibility is "expect" it should not be added to the pages list.
+        if ($visibility) {
+          $pages .= "\r\nsearch-blank";
+        }
+
+        // Compare the lowercase internal and lowercase path alias (if any).
+        $page_match = drupal_match_path($path, $pages);
+        if ($path != $_GET['q']) {
+          $page_match = $page_match || drupal_match_path($_GET['q'], $pages);
+        }
```

Comment:
Less copy/paste - more filtering.

Reply:
?

Reply:
My comment refers to the situation above.